### ISSUE TITLE
Lazy() support required for recursive schemas

### ISSUE THREAD

**Issue by laurence-myers** [joiful-ts/joiful#10]
E.g. consider a tree node which has children, each child can be a tree node. Passing the class to "@NestedArray()", or calling "@ArrayConstraints.Items(getJoiSchema(MySchemaClass))", does not seem to work. I think this is because the schema has not been fully defined by the time it's passed to the decorator, but these decorators force the resolution of the (incomplete) Joi schema.

e.g.
```typescript
class TreeNode {
  @Required()
  tagName: string;

  @NestedArray(TreeNode)
  children: TreeNode[];
}
```

The "children" property will get stripped out.

This can be resolved by adding support for Joi's "lazy()" function, which allows passing in a function that returns a schema. (In our case, we could make it also accept a schema class.)

https://github.com/hapijs/joi/issues/379

**Comment by laurence-myers** [comment]
I've addressed this in commit 0066a459b9935f33d542b850d8048f873f0ea36a.

I think the syntax could be improved. Currently, the callback must return a Joi schema. This means you need to call `getJoiSchema()`. If the property is an array, you also need to wrap that in `Joi.array().items()`. This ends up looking quite verbose, as in the example below.

```
class TreeNode {
    @Required()
    tagName! : string;

    @Lazy(() => Joi.array().items(getJoiSchema(TreeNode)))
    children! : TreeNode[];
}
```

Maybe there should be special `@LazyNested()` and `@LazyNestedArray()` decorators?

**Comment by laurence-myers** [comment]
Fixed in [v0.0.12](https://github.com/laurence-myers/tsdv-joi/releases/tag/v0.0.12).

Refinements to this decorator's API can be made in the future, PRs are welcome.
